### PR TITLE
don't succeed `evaluate_obligation` query if new opaque types were registered

### LINKED ISSUE #98604: Broken Mir Resulting From Async Function Pointer Cast
### Code

```Rust
type AsyncFnPtr = Box<
    dyn Fn() -> std::pin::Pin<Box<dyn std::future::Future<Output = ()>>>,
>;

async fn test() {}

#[allow(unused_must_use)]
fn main() {
    Box::new(test) as AsyncFnPtr;
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.63.0-nightly (fdca237d5 2022-06-24)
binary: rustc
commit-hash: fdca237d5194bf8a1c9b437ebd2114d1c2ba6195
commit-date: 2022-06-24
host: x86_64-apple-darwin
release: 1.63.0-nightly
LLVM version: 14.0.5
```

### Error output

```
error: internal compiler error: no errors encountered even though `delay_span_bug` issued

error: internal compiler error: broken MIR in DefId(0:7 ~ playground[5af9]::main) (NoSolution): could not prove Binder(TraitPredicate(<std::boxed::Box<fn() -> impl std::future::Future<Output = ()> {test}> as std::ops::CoerceUnsized<std::boxed::Box<dyn std::ops::Fn() -> std::pin::Pin<std::boxed::Box<dyn std::future::Future<Output = ()>>>>>>, polarity:Positive), [])
 --> src/main.rs:9:5
  |
9 |     Box::new(test) as AsyncFnPtr;
  |     ^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_borrowck/src/type_check/canonical.rs:151:13

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1425:13
stack backtrace:
   0:     0x7fd68ac845dd - std::backtrace_rs::backtrace::libunwind::trace::ha6648b7014364bc2
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7fd68ac845dd - std::backtrace_rs::backtrace::trace_unsynchronized::h12a4ddf96a4e2ff5
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7fd68ac845dd - std::sys_common::backtrace::_print_fmt::hf87c549bd1cc89dc
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7fd68ac845dd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4ffedd9ce9224034
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7fd68acdd91c - core::fmt::write::hf124b4c1c8ae0d2b
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/core/src/fmt/mod.rs:1198:17
   5:     0x7fd68ac75c51 - std::io::Write::write_fmt::hbe26d06d41c58912
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/io/mod.rs:1672:15
   6:     0x7fd68ac872a5 - std::sys_common::backtrace::_print::h6d5a85ca9b3358b3
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7fd68ac872a5 - std::sys_common::backtrace::print::h7666fc3156204d9f
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7fd68ac872a5 - std::panicking::default_hook::{{closure}}::hfa94c116b7086896
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/panicking.rs:295:22
   9:     0x7fd68ac86fc6 - std::panicking::default_hook::hc8deb5a55c834c81
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/panicking.rs:314:9
  10:     0x7fd68b481524 - rustc_driver[3a72345c5a3a9278]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7fd68ac8797a - std::panicking::rust_panic_with_hook::h16757dfa430f6014
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/panicking.rs:702:17
  12:     0x7fd68c63a111 - std[a5eff6be7f10e3b3]::panicking::begin_panic::<rustc_errors[8af9c38e420da921]::ExplicitBug>::{closure#0}
  13:     0x7fd68c637cf6 - std[a5eff6be7f10e3b3]::sys_common::backtrace::__rust_end_short_backtrace::<std[a5eff6be7f10e3b3]::panicking::begin_panic<rustc_errors[8af9c38e420da921]::ExplicitBug>::{closure#0}, !>
  14:     0x7fd68c6545f6 - std[a5eff6be7f10e3b3]::panicking::begin_panic::<rustc_errors[8af9c38e420da921]::ExplicitBug>
  15:     0x7fd68c651596 - std[a5eff6be7f10e3b3]::panic::panic_any::<rustc_errors[8af9c38e420da921]::ExplicitBug>
  16:     0x7fd68df2266c - <rustc_errors[8af9c38e420da921]::HandlerInner as core[993276dbfd88d183]::ops::drop::Drop>::drop
  17:     0x7fd68d64d488 - core[993276dbfd88d183]::ptr::drop_in_place::<rustc_session[ba178145ab1da357]::parse::ParseSess>
  18:     0x7fd68d64fda3 - <alloc[40aff515e5175a27]::rc::Rc<rustc_session[ba178145ab1da357]::session::Session> as core[993276dbfd88d183]::ops::drop::Drop>::drop
  19:     0x7fd68d6763bd - core[993276dbfd88d183]::ptr::drop_in_place::<rustc_interface[48a3913523791f79]::interface::Compiler>
  20:     0x7fd68d675da4 - rustc_span[a7833de75ef67468]::with_source_map::<core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>, rustc_interface[48a3913523791f79]::interface::create_compiler_and_run<core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>, rustc_driver[3a72345c5a3a9278]::run_compiler::{closure#1}>::{closure#1}>
  21:     0x7fd68d652532 - <scoped_tls[5373c8190146815]::ScopedKey<rustc_span[a7833de75ef67468]::SessionGlobals>>::set::<rustc_interface[48a3913523791f79]::interface::run_compiler<core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>, rustc_driver[3a72345c5a3a9278]::run_compiler::{closure#1}>::{closure#0}, core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>>
  22:     0x7fd68d66751f - std[a5eff6be7f10e3b3]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[48a3913523791f79]::util::run_in_thread_pool_with_globals<rustc_interface[48a3913523791f79]::interface::run_compiler<core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>, rustc_driver[3a72345c5a3a9278]::run_compiler::{closure#1}>::{closure#0}, core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>>::{closure#0}, core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>>
  23:     0x7fd68d667659 - <<std[a5eff6be7f10e3b3]::thread::Builder>::spawn_unchecked_<rustc_interface[48a3913523791f79]::util::run_in_thread_pool_with_globals<rustc_interface[48a3913523791f79]::interface::run_compiler<core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>, rustc_driver[3a72345c5a3a9278]::run_compiler::{closure#1}>::{closure#0}, core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>>::{closure#0}, core[993276dbfd88d183]::result::Result<(), rustc_errors[8af9c38e420da921]::ErrorGuaranteed>>::{closure#1} as core[993276dbfd88d183]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  24:     0x7fd68ac914f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h63ee45655a45e444
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/alloc/src/boxed.rs:1951:9
  25:     0x7fd68ac914f3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha9e4bd17f57e4e2f
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/alloc/src/boxed.rs:1951:9
  26:     0x7fd68ac914f3 - std::sys::unix::thread::Thread::new::thread_start::h4bd37e41a2fb87e0
                               at /rustc/2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f/library/std/src/sys/unix/thread.rs:108:17
  27:     0x7fd68abb4609 - start_thread
  28:     0x7fd68aad7133 - clone
  29:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.64.0-nightly (2f3ddd9f5 2022-06-27) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type bin -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   3: core::ptr::drop_in_place::<rustc_session::parse::ParseSess>
   4: <alloc::rc::Rc<rustc_session::session::Session> as core::ops::drop::Drop>::drop
   5: core::ptr::drop_in_place::<rustc_interface::interface::Compiler>
   6: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
   7: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
```

</p>
</details>

### LINKED ISSUE #98608: Function type with opaque return type can be coerced to `dyn Fn` with arbitrary return type
I tried this code:

```rust
fn hi() -> impl Sized { std::ptr::null::<u8>() }

fn main() {
    let b: Box<dyn Fn() -> Box<u8>> = Box::new(hi);
    let boxed = b();
    let null = *boxed;
    println!("{null:?}");
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a78a1c65a7788a1fc16bbb49823e31b1

I expected to see this happen: Compiler error, since `impl Sized` cannot be coerced to `Box<u8>`.

Instead, this happened: Code compiles, segfault

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (2f3ddd9f5 2022-06-27)
binary: rustc
commit-hash: 2f3ddd9f594adf9773547aa7cedb43c4ac8ffd2f
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #98608
fixes #98604

The root cause of all this is that in type flag computation we entirely ignore nongeneric things like struct fields and the signature of function items. So if a flag had to be set for a struct if it is set for a field, that will only happen if the field is generic, as only the generic parameters are checked.

I now believe we cannot use type flags to handle opaque types. They seem like the wrong tool for this. 

Instead, this PR replaces the previous logic by adding a new variant of `EvaluatedToOk`: `EvaluatedToOkModuloOpaqueTypes`, which says that there were some opaque types that got hidden types bound, but that binding may not have been legal (because we don't know if the opaque type was in its defining scope or not).

### CODE REVIEW

**Comment by compiler-errors** [comment]
Oof

**Comment by oli-obk** [comment]
@bors try @rust-timer queue

while this has a real impact in secondary benchmarks, the primary ones could be due to other commits, so I've only pushed the actual soundness fix commit

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
If there is a clean way to mitigate the perf impact, it would be great to do that but I think it's acceptable to merge as-is to resolve the unsoundness.

**Comment by oli-obk** [comment]
Nominating for compiler team discussion. Are we ok with significantly regressing performance with a small targetted soundness fix? There'll happen a proper resolution of the bug aferwards that will get us perf back, but it's a significantly larger refactoring that may take a few months gov et all parts of merged

**Comment by compiler-errors** [comment]
i don't see it on the perf queue...

@rust-timer queue

**Comment by compiler-errors** [comment]
uh, wait how do you queue something that already has CI artifacts

**Comment by lqd** [comment]
maybe:
@rust-timer build 65e0f3e181384cf41c894d47ff1503f3b839b82e

**Comment by compiler-errors** [comment]
rust-timer seems to be having a bad day

**Comment by compiler-errors** [comment]
@rust-timer build 65e0f3e181384cf41c894d47ff1503f3b839b82e

**Review by compiler-errors — APPROVED** [review]
i greatly prefer this approach, and the perf hit is totally fine imo (or more likely noise tbh)

**Comment by oli-obk** [comment]
@bors try @rust-timer queue

one more cleanup that is now possible due to the latest version.

**Comment by oli-obk** [comment]
Nominating for beta backport,  as the diff isn't terrible and it fixes a soundness bug.

In case a minor stable release happens, this should also be considered for inclusion

File: compiler/rustc_middle/src/traits/select.rs [hunk 1]
```diff
@@ -174,6 +174,10 @@ pub enum SelectionCandidate<'tcx> {
 pub enum EvaluationResult {
     /// Evaluation successful.
     EvaluatedToOk,
+    /// Evaluation successful, but need to rerun because opaque types got
+    /// hidden types assigned without it being known whether the opaque types
+    /// are within their defining scope
+    EvaluatedToOkModuloOpaqueTypes,
```
Comment (wesleywiser) [inline compiler/rustc_middle/src/traits/select.rs:180]: Is `must_apply_modulo_regions()`'s implementation still correct? It looks like that will return `true` for `EvaluatedToOkModuloOpaqueTypes`; is that valid?

Comment (oli-obk) [reply compiler/rustc_middle/src/traits/select.rs:182]: yea, it's not, but I also wasn't able to create a test that actually was a problem. Just to be safe, I moved the opaque types down

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
@bors r+ p=1

Fixes P-critical issue.

**Pull request merged**